ethers-etherscan/src/contract.rs

```diff
@@ -347,14 +422,38 @@ impl Client {
 
         let query =
             self.create_query("contract", "getsourcecode", HashMap::from([("address", address)]));
-        let response: Response<Vec<Metadata>> = self.get_json(&query).await?;
-        if response.result.iter().any(|item| item.abi == "Contract source code not verified") {
+        let mut res = self.get(&query).await?;
+
+        // Source code is not verified
+        if res.contains("Contract source code not verified") {
             if let Some(ref cache) = self.cache {
                 cache.set_source(address, None);
             }
             return Err(EtherscanError::ContractCodeNotVerified(address))
         }
-        let res = ContractMetadata { items: response.result };
+
+        // Etherscan can either return one raw string that includes all of the source code for a
+        // verified contract or a [SourceCodeMetadata] struct surrounded in an extra set of {}.
+        // {"SourceCode": "{{}}", ..} -> {"SourceCode": {}, ..}
+        let start = r#""SourceCode":"{{"#;
+        let end = r#"}}","ABI""#;
+        if let Some(start_idx) = res.find(start) {
+            // this should not fail
+            let end_idx = res
+                .find(end)
+                .ok_or_else(|| EtherscanError::Unknown(format!("Malformed response {}", res)))?;
+            // the SourceCode string value
+            let range = start_idx + 13..end_idx + 3;
+            let source_code = &res[range.clone()];
+            // parse the escaped characters
+            let parsed: String = serde_json::from_str(source_code)?;
+            // skip the first `{` and last `}`
+            let source_code = &parsed[1..parsed.len() - 1];
+            // replace
+            res.replace_range(range, source_code);
+        }
```

Comment:
can we use untagged enum here instead?

Reply:
No the field is always a string, and have to check manually if it starts with `{{` / ends with `}}`. It's a bit hacky but couldn't figure out another way

Reply:
it's not obvious to me how the two variants differ.

1. the value of `SourceCode ` is a string
2. the value of `SourceCode ` is a json object
?

then let's just create a new type with a source_code field and impl deserialize manually instead of hacking with indices here

Reply:
response.result[n].SourceCode is either:
- a string containing all of the source code
- a string containing a JSON object wrapped in an extra set of `{}`
not sure how to implement `Deserialize` manually since in both cases it's a string; we always have to parse it as that first, then check for a JSON, then deserialize that one too

Reply:
I'd just use an untagged enum here for convenience, where the json variant is the first attempt.